explainer.md

```diff
@@ -215,6 +243,162 @@ third-party then signs either the exchanges in a package or the package as a
+as prefetches. If those point at signed exchanges, this process repeats.
+
+### Caching the signed response
+
+If the signed exchange was requested as a navigation or subresource (i.e. *not*
+prefetches), the client tries to cache the request→response pair it contains.
+
+This *doesn't* happen if:
+
+* The signed exchange's request headers aren't sufficiently similar (TBD) to the
+  request headers the client would use for a normal request in the same context.
+  This prevents a malicious intermediate from sticking the wrong
+  content-negotiated resource in the HTTP cache.
+* There's a response in the HTTP (or preload?) cache with a newer Date header
+  than the signed exchange's response. This prevents some downgrade attacks.
+
+In either of these cases, the client just skips to the redirect in the next
+step.
+
+If we're still here, the signed exchange is put into the [preload
+cache](https://github.com/whatwg/fetch/issues/590) and, if the response headers
+allow it (and review of this explainer indicates it's a good idea), the [HTTP
+cache](https://tools.ietf.org/html/rfc7234).
+
+If we put the signed exchange in the HTTP cache, its freshness has to be bounded
+by the shorter of the normal HTTP cache lifetime or the signature's expiration.
+For *later* loads of the logical URL (in particular, not the load that's
+happening through the signed exchange, since it's fulfilled using the preload
+cache), a stale entry can be revalidated in the following ways:
+
+* If the `Signature` is expired but the HTTP caching information is fresh, the
+  client can fetch the `validityUrl `to update just the signature. It *must not*
+  send an `If-None-Match` or `If-Modified-Since` request to update the cache,
+  because the `Signature` expiring means we don't trust the claimed ETag or date
+  anymore.
+* If the `Signature` is valid but the HTTP caching information is stale, the
+  client can send the logical URL an `If-None-Match` or `If-Modified-Since`
+  request hoping for a 304. Note that the client has to keep the original
+  response headers if it intends to use the `validityUrl` to update the
+  signature in the future.
+* If neither is valid, the client could first update the signature and then
+  check for a 304 (or even do both concurrently), but it may be easier to just
+  do an unconditional request.
+
+### Navigations and subresources redirect
+
+At this point the client redirects to the signed exchange's logical URL. For
+navigations, this request goes through the logical URL's Service Worker. For
+subresources, the embedder's Service Worker already handled the request when it
+chose to return a signed exchange, so we currently think it doesn't need a
+second chance.
+
+If the request goes through a Service Worker (and caching wasn't skipped above),
+the `FetchEvent` needs to include some notification that there's a response
+available in the preload cache. We currently think the
```

Comment:
I don't think the `FetchEvent` *needs* to include the notification, but if SW wants to know about that we can do that by utilizing navigation preload (only for the SWs that have enabled the feature), is the argument.

Reply:
I agree that we can probably ship without this notification, but I think people will want it sooner or later. I'll downgrade my language here.

The existing navigation preload only applies to navigations, while signed exchanges also offer a response for subresources, so I think we'll need to extend its current spec. I also don't think we need opt-in to expose this: navigation preload needs opt-in because it adds an extra maybe-useless request, but this just exposes something that's happening anyway.